design.md

```diff
@@ -1,2 +1,97 @@
+
+## Server
+
+The server is responsible for authentication and authorization before handing off the main work to the load balancing library. Authentication will be done with mTLS, with certificates stored in the repo for the sake of simplicity. Authorization will also be hard coded; we'll simply keep a struct of which upstreams are available to which clients.
+
+The server will accept TCP connections on a dedicated port for each upstream - e.g. port 7777 for application A, port 8888 for application B. Whenever it receives a connection, it will spin up a goroutine to handle it like so:
+
+- Read the client's identity from the SAN in the cert (e.g. `admin`, `user1`, `pikachu` - again, hard coded to keep things simple)
+- Check the client identity against the hard-coded authorization data
+    - If the client doesn't have access to the upstream for the port it's connected on, send it an error message, close its connection and return without further work
+- Ask the rate limiter whether the client is allowed to make a request right now
+    - If the client is rate limited, send it an error message, close its connection and return without further work
+- Forward the client's connection to the load balancing library to pass it to an appropriate host for the upstream (or send it an error and close it if there's a problem connecting to an upstream)
+
+## User Experience / Testing
+
+Clients would be able to communicate with this load balancer the way they would for any TLS-secured TCP traffic. To simplify the review process and add test coverage, I'll write end-to-end tests for the application in Go.
+
+The following scenarios are ranked in what I think are the highest priority. Depending on time constraints and reviewer interest, the last on the list may not be necessary.
+
+1. Connecting to the load balancer on a port for upstream A correctly routes you to a host for upstream A (the upstream will just be some dummy TCP listeners created in the test)
+2. Connecting to the load balancer on a port corresponding to an upstream the client isn't authorized for results in an error
+3. If upstream A has a host with 1 active connection, another request to upstream A will route to a second host with 0 active connections
+4. Making many paralell requests from the same client results in a rate-limited rejection (timing will make this hard to reproduce perfectly; will probably just throw a bunch of goroutines at it and assert at least one of them got the rate limiting error message)
+5. Same as #1, but with a "hello world" HTTP upstream instead of a raw TCP listener
+
+## Other Considerations
+
+- For this exercise, we'll handle errors by simply writing an error message to the client's TCP connection and closing it. In a production system we'd want something more robust - perhaps custom, documented error codes, for example.
```

Comment:
What is the error message? A UTF-8 string? JSON?

Reply:
I was thinking a raw UTF-8 string for simplicity's sake, as I anticipate just writing strings from the upstreams too. (e.g. `Hello from upstream A, host 1`) But we could certainly do JSON if we wanted. I'll go forward with strings for now but am certainl open to changing.